include/boost/windows_sspi/detail/sspi_impl.hpp

```diff
@@ -0,0 +1,126 @@
+//
+// windows_sspi/detail/sspi_impl.hpp
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+//
+// Copyright (c) 2020 Kasper Laudrup (laudrup at stacktrace dot dk)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BOOST_WINDOWS_SSPI_DETAIL_SSPI_IMPL_HPP
+#define BOOST_WINDOWS_SSPI_DETAIL_SSPI_IMPL_HPP
+
+#include <boost/windows_sspi/detail/sspi_functions.hpp>
+
+#include <boost/core/ignore_unused.hpp>
+
+namespace boost {
+namespace windows_sspi {
+namespace detail {
+
+namespace net = boost::asio;
+
+class sspi_impl {
+public:
+  sspi_impl(CtxtHandle* context)
+    : m_context(context) {
+  }
+
+  // TODO: Calculate this once when handshake is complete
+  SecPkgContext_StreamSizes stream_sizes() const {
+    SecPkgContext_StreamSizes stream_sizes;
+    SECURITY_STATUS sc = detail::sspi_functions::QueryContextAttributes(m_context, SECPKG_ATTR_STREAM_SIZES, &stream_sizes);
+
+    // TODO: Signal error to user (throw exception or use error code?)
```

Comment:
The most likely way that this would fail would be if the handshake hasn't been completed, which would most likely give a rather cryptic error message.

It would most likely be better to ensure that calling any of the read or write functions would fail with a useful error if that was the case, but for now I'll refactor the code to ensure that an error is returned to the user. It is required if the function would fail for any other reason anyway.